app/helpers/lti_helper.rb

```diff
@@ -0,0 +1,151 @@
+      lti_user.update!(lti_user_id: lms_user[:lti_user_id])
+    end
+    error
+  end
+
+  def grade_sync(lti_deployment, assessment)
+    scopes = [LtiDeployment::LTI_SCOPES[:score], LtiDeployment::LTI_SCOPES[:results]]
+    create_or_update_lti_assessment(lti_deployment, assessment)
+    auth_data = lti_deployment.lti_client.get_oauth_token(scopes)
+    line_item = lti_deployment.lti_line_items.find_by!(assessment: assessment)
+    curr_results = get_current_results(line_item, auth_data, scopes)
+    score_uri = URI("#{line_item.lti_line_item_id}/scores")
+    req = Net::HTTP::Post.new(score_uri)
+
+    if assessment.is_a?(Assignment)
+      marks = get_assignment_marks(lti_deployment, assessment)
+    else
+      marks = get_grade_entry_form_marks(lti_deployment, assessment)
+    end
+    if marks.empty?
+      raise I18n.t('lti.no_grades')
+    end
+    marks.each do |lti_user_id, mark|
+      # Only send if the mark has not been previously sent to the LMS
+      # or if the mark differs from the LMS mark.
+      marked_by_lms = curr_results.find { |result| result['userId'] == lti_user_id }
+      if marked_by_lms.nil? || marked_by_lms['resultScore'] != mark
+        payload = {
+          timestamp: Time.current.iso8601,
+          scoreGiven: mark,
+          scoreMaximum: assessment.max_mark.to_f,
+          activityProgress: 'Completed',
+          gradingProgress: 'FullyGraded',
+          userId: lti_user_id
+        }
+        req.set_form_data(payload)
+        lti_deployment.send_lti_request!(req, score_uri, auth_data, scopes)
+      end
+    end
+  end
+
+  # Returns a hash mapping lti_user_id to marks
+  # for each released mark where the user has an lti_user_id
+  def get_assignment_marks(lti_deployment, assignment)
+    released_marks = assignment.released_marks
+                               .joins(grouping: [{ group: :student_memberships }])
+                               .merge(StudentMembership.joins(role: [{ user: :lti_users }]))
+                               .where('lti_users.lti_client': lti_deployment.lti_client)
+                               .pluck('lti_users.lti_user_id', 'results.id')
+    result_ids = released_marks.map { |mark| mark[1] }
+    grades = Result.get_total_marks(result_ids)
+    grades.map do |grade|
+      student = released_marks.find { |g| g[1] == grade[0] }
+      next if student.nil?
+      [student[0], grade[1]]
+    end.to_h
+  end
+
+  # Returns a hash mapping lti_user_id to marks
+  # for each released mark where the user has an lti_user_id
+  def get_grade_entry_form_marks(lti_deployment, grade_entry_form)
+    student_data = grade_entry_form.released_marks
+                                   .joins(role: :user)
+                                   .joins(user: :lti_users)
+                                   .where('lti_users.lti_client': lti_deployment.lti_client,
+                                          'grade_entry_students.released_to_student': true)
```

Comment:
This shouldn't be necessary, the `released_marks` association should generate this condition already (but if it doesn't, that's another problem to address).